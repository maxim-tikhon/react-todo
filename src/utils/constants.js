export const FILTER = {
	ALL: 'all',
	ACTIVE: 'active',
	COMPLETED: 'completed',
};

export const TODO_STOARGE_NAME = 'tasks';
