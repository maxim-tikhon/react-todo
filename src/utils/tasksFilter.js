const taskFilter = {
	filterTasks(tasks, filter) {
		switch (filter) {
			case "all":
				return tasks;
			case "active":
				return tasks?.filter((task) => !task.completed);
			case "completed":
				return tasks?.filter((task) => task.completed);
			default:
				return tasks;
		}
	}
};

export default taskFilter;
