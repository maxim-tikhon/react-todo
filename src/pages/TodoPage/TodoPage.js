import TodoTemplate from "../../templates/TodoTemplate/TodoTemplate";
import { useSelector, useDispatch } from "react-redux";
import {
	addTask,
	deleteTask,
	checkTask,
	changeFilter,
	clearCompletedTasks,
	loadTasks
} from "../../store/slices/todoSlice";
import { useEffect } from "react";
import { TODO_STOARGE_NAME } from "../../utils/constants";

const TodoPage = () => {
	const tasks = useSelector((state) => state.todo.tasks);
	const filter = useSelector((state) => state.todo.filter);

	const dispatch = useDispatch();

	useEffect(() => {
		const storedTasks = localStorage.getItem(TODO_STOARGE_NAME);

		if (storedTasks) {
			const tasks = JSON.parse(storedTasks);
			dispatch(loadTasks(tasks));
		}
	}, []);

	return (
		<TodoTemplate
			tasks={tasks}
			filter={filter}
			onAddTask={(task) => dispatch(addTask(task))}
			onDeleteTask={(id) => dispatch(deleteTask(id))}
			onCheckTask={(task) => dispatch(checkTask(task))}
			onChangeFilter={(filter) => dispatch(changeFilter(filter))}
			onClearCompletedTasks={() => dispatch(clearCompletedTasks())}
		/>
	);
};

export default TodoPage;
