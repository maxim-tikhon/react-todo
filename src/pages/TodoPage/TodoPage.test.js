import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { fireEvent, render, screen } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import TodoPage from './TodoPage';
import { FILTER } from '../../utils/constants';
import todoReducer from '../../store/slices/todoSlice';

describe('TodoPage', () => {
	let store;

	beforeEach(() => {
		store = configureStore({
			reducer: {
				todo: todoReducer,
			},
			preloadedState: {
				todo: {
					tasks: [
						{
							id: '1',
							description: 'Task',
							completed: false,
						},
						{
							id: '2',
							description: 'Task 2',
							completed: true,
						},
					],
					filter: FILTER.ALL,
				},
			},
		});
	});

	afterEach(() => {
		store = null;
	});

	it('renders correctly', () => {
		const output = renderer
			.create(
				<Provider store={store}>
					<TodoPage />
				</Provider>
			)
			.toJSON();
		expect(output).toMatchSnapshot();
	});

	it('dispatches addTask action when new task entered', () => {
		const { container } = render(
			<Provider store={store}>
				<TodoPage />
			</Provider>
		);

		const inputElement = container.querySelector('input[type="text"]');
		fireEvent.change(inputElement, { target: { value: 'New Task' } });
		fireEvent.keyDown(inputElement, { key: 'Enter', code: 'Enter' });

		const state = store.getState();
		expect(state.todo.tasks[2].description).toEqual('New Task');
	});

	it('dispatches changeFilter action when filter is changed', () => {
		render(
			<Provider store={store}>
				<TodoPage />
			</Provider>
		);

		const filter = FILTER.COMPLETED;
		fireEvent.click(screen.getByText('Completed'));

		const state = store.getState();
		expect(state.todo.filter).toBe(filter);
	});

	it('dispatches clearCompletedTasks action when "Clear Completed" is clicked', () => {
		render(
			<Provider store={store}>
				<TodoPage />
			</Provider>
		);

		fireEvent.click(screen.getByText('Clear Completed'));

		const state = store.getState();
		expect(state.todo.tasks.length).toBe(1);
	});
});
