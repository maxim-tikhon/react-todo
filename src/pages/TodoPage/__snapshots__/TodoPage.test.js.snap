// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPage renders correctly 1`] = `
<div
  className="todoContainer"
>
  <section
    className="todoSection"
  >
    <header>
      <h1
        className="todoHeader"
      >
        TODO
      </h1>
    </header>
    <div
      className="card single"
    >
      <input
        className="customCheckbox"
        disabled={true}
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="label"
      />
      <input
        className="textInput newTaskInput"
        onKeyDown={[Function]}
        placeholder="Please enter a new task and click Enter to save"
        type="text"
      />
    </div>
    <div
      className="todoList"
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled={false}
        aria-roledescription="sortable"
        className="card"
        onPointerDown={[Function]}
        role="button"
        style={
          Object {
            "transform": undefined,
            "transition": undefined,
          }
        }
        tabIndex={0}
      >
        <input
          checked={false}
          className="customCheckbox"
          id="Task"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="label"
          htmlFor="Task"
        >
          Task
        </label>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium deleteIcon css-i4bv87-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled={false}
        aria-roledescription="sortable"
        className="card"
        onPointerDown={[Function]}
        role="button"
        style={
          Object {
            "transform": undefined,
            "transition": undefined,
          }
        }
        tabIndex={0}
      >
        <input
          checked={true}
          className="customCheckbox"
          id="Task 2"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="label"
          htmlFor="Task 2"
        >
          Task 2
        </label>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium deleteIcon css-i4bv87-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
      <div
        id="DndDescribedBy-0"
        style={
          Object {
            "display": "none",
          }
        }
      >
        
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
      </div>
      <div
        aria-atomic={true}
        aria-live="assertive"
        id="DndLiveRegion-0"
        role="status"
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(100%)",
            "height": 1,
            "margin": -1,
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      />
      <div
        className="card"
      >
        <div
          className="todoFooter"
        >
          <span
            className="totalCount"
          >
            1
             
            task
             left
          </span>
          <div
            className="buttonsContainer"
          >
            <button
              className="linkButton active"
              onClick={[Function]}
              type="button"
            >
              All
            </button>
            <button
              className="linkButton"
              onClick={[Function]}
              type="button"
            >
              Active
            </button>
            <button
              className="linkButton"
              onClick={[Function]}
              type="button"
            >
              Completed
            </button>
          </div>
          <button
            className="linkButton"
            onClick={[Function]}
            type="button"
          >
            Clear Completed
          </button>
        </div>
      </div>
    </div>
    <div
      className="todoNote"
    >
      Drag and drop to reoder list
    </div>
  </section>
</div>
`;
